### PR TITLE
Konrad/x509 field

### LINKED ISSUE #75: Rename X509 Credential field to `chain` and change structure (#408)
https://github.com/mlswg/mls-protocol/pull/408

### PR DESCRIPTION
**Pull request by kkohbrok**
Fixes #75 and thus partially #21.

More concretely, it adds a `Credential` struct to match that in the spec as well as an `MLSCredentialType` enum for the individual credential types. The `Credential` impl gets an additional `identity` getter. Also, the `Certificate` credential gets its own struct, although the corresponding implementations panic at the moment, since we don't yet support `X509` certs.

### CODE REVIEW

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Not sure what this is doing. Please add a description to the PR.

File: src/creds.rs [hunk 1]
```diff
@@ -105,30 +106,66 @@ impl Codec for CredentialType {
...
             ),
+            // TODO: implement verification for X509 certificates
+            MLSCredentialType::X509(_) => panic!("X509 certificates are not yet implemented."),
+        }
+    }
+    /// Get the identity of a given credential.
+    pub fn get_identity(&self) -> &Vec<u8> {
+        match &self.credential {
+            MLSCredentialType::Basic(basic_credential) => &basic_credential.identity,
+            // TODO: implement getter for identity for X509 certificates
```
Comment (franziskuskiefer) [inline src/creds.rs:146]: A TODO should always get an issue referenced if possible. We should have an issue that is for implementing x509 support.

File: src/creds.rs [hunk 2]
```diff
@@ -105,30 +106,66 @@ impl Codec for CredentialType {
...
 impl Codec for Credential {
     fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
-        match self {
-            Credential::Basic(basic_credential) => {
+        match &self.credential {
+            MLSCredentialType::Basic(basic_credential) => {
                 CredentialType::Basic.encode(buffer)?;
                 basic_credential.encode(buffer)?;
             }
+            // TODO: implement encoding for X509 certificates
```
Comment (franziskuskiefer) [inline src/creds.rs:167]: Same here.

**Review by franziskuskiefer — COMMENTED** [review]
So the structure of credentials must change. I added some suggestions to this. Because this code basically works we can leave it as is and I'll rewrite it in #74, or you can try to make credentials rusty.

I started on #74, which will touch some of these things as well and there's some stuff missing/that needs changing. I'll hold off on that until this one gets done.

File: src/creds.rs [hunk 3]
```diff
@@ -77,6 +77,7 @@ impl Codec for Identity {
     // }
 }
 
+/// Enum for Credential Types. We only need this for encoding/decoding.
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply src/creds.rs:80]: Yes, but there is no need for our nice rust structs to look like ugly C structs. So I thought in the `Credentials` struct, we can use an enum that contains the actual data as opposed to a `u16` enum plus something that contains whatever data is specified in that enum.

File: src/creds.rs [hunk 4]
```diff
@@ -105,30 +106,66 @@ impl Codec for CredentialType {
...
+#[derive(Debug, PartialEq, Clone)]
+pub enum MLSCredentialType {
     Basic(BasicCredential),
+    X509(Certificate),
+}
+
+/// Struct containing MLS credential data, where the data depends on the type.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Credential {
+    credential: MLSCredentialType,
```
[the remark this answers is not in the document]
Comment (kkohbrok) [reply src/creds.rs:125]: It does. Note the difference between `MLSCredentialType` and `CredentialType`. It's similar with the `ContentType` in the `MLSPlaintext`, I think.

**Review by franziskuskiefer — DISMISSED** [review]
Lets merge and tackle the rewrite together.

**Review by franziskuskiefer — APPROVED** [review]

**Pull request merged**